obs: add otel tracing

artem-tkachuk opened the pull request:
# Add OpenTelemetry tracing to Spegel

## Summary
- Add OpenTelemetry tracing to registry, metrics, and OCI HTTP flows
- Centralize HTTP instrumentation in `httpx` using `otelhttp`
- Respect existing OTEL providers/propagators when already configured
- Expose OTEL config in the Helm chart and wire flags in the DaemonSet
- Add unit tests covering OTEL setup, sampler behavior, and HTTP propagation

## Why
Provide request‑level tracing for debugging and observability without breaking embedders that already configure OTEL.

## Details
- Instrumentation:
  - HTTP handlers (registry, metrics) via `httpx.WrapHandler`
  - HTTP client transport (OCI requests) via `httpx.WrapTransport`
  - P2P operations (lookup/bootstrap)
- Helm: `values.yaml` exposes `.Values.spegel.otel.{endpoint,insecure,serviceName,sampler}` and flags are wired into the DaemonSet.
- OTEL setup avoids overriding global tracer provider/propagator when already set.
  
## Test plan
- `go test ./...`
- `golangci-lint run ./...`

## Impact
- OTEL is always built in; if a collector is not reachable, exporting may log warnings but does not fail requests.

## Screenshots
<img width="1904" height="1264" alt="Screenshot 2026-01-22 at 19 25 57" src="https://github.com/user-attachments/assets/f0354f65-a4e9-46fb-a5d6-50c35df470c1" />

<img width="2012" height="799" alt="Screenshot 2026-01-22 at 19 25 32" src="https://github.com/user-attachments/assets/d4b254a2-daa2-4040-8f46-2bdfa4b73815" />

## Follow-up
- I am working on a separate PR to add the k8s itest tracing setup and document the repro steps.

phillebaba commented on internal/otelx/otelx_noop.go:
```diff
@@ -0,0 +1,51 @@
+//go:build !otel
```
This should not be needed once OTEL is not an optional feature.

artem-tkachuk replied:
Removed in 7b5cb69eedb317e99366e706dfc7085981791d67

phillebaba reviewed (changes requested):
@artem-tkachuk as we discussed during last weeks community meeting. We should remove the optional build tags and then also the docker compose tests. Keep things slim with adding tracing, and then we can add more features as needed. Once you remove the things that are not needed it will reduce the noise and make it easier to review.

What would be nice is to get to a first point where we are setting up otelhttp so that both trace headers are forwarded by the clients and also registered for requests. Does that sound like a good idea?

brandond commented on main.go:
```diff
@@ -217,7 +250,7 @@ func registryCommand(ctx context.Context, args *RegistryCmd) error {
 	regSrv := &http.Server{
 		Addr:    args.RegistryAddr,
-		Handler: reg.Handler(log),
+		Handler: otelx.WrapHandler("registry", reg.Handler(log)),
```
Also keep in mind that embedding projects may have their own otel config that this will need to play nice with.

artem-tkachuk replied:
HTTP instrumentation now uses `otelhttp` and lives in `httpx` (handlers + transport), and OTEL setup skips overriding existing tracer providers/propagators if already configured 15df91fa380f16af03ebe45fb9caf8c382555745

artem-tkachuk committed:
deps: add otel modules and tidy itest deps

Add OpenTelemetry module requirements and refresh integration test module
files while keeping golangci-lint import aliasing consistent.

artem-tkachuk committed:
otelx: add OTEL setup and helpers

Initialize OTEL tracing with env-backed defaults and provide helper
functions for span creation and log enrichment.

artem-tkachuk committed:
httpx: instrument registry and OCI requests

Centralize otelhttp wrapping for handlers and transports, then route
registry and OCI requests through the new helpers.

artem-tkachuk committed:
main: wire OTEL config and p2p spans

Plumb OTEL config from flags into setup and add spans around P2P
bootstrap and lookup operations.

artem-tkachuk committed:
otelx: add setup behavior tests

Cover override vs. reuse behavior, propagator handling, and sampling
defaults for the OTEL setup flow.

artem-tkachuk committed:
charts: expose OTEL values in daemonset

Add OTEL chart values and render them into the daemonset args with a
guard for empty endpoints, then refresh helm docs.

artem-tkachuk committed:
docs: update README for OTEL

phillebaba commented on main_test.go:
```diff
@@ -0,0 +1,27 @@
+package main
```
Remove this file. I am working on getting coverage reports from the Kuberentes integrationtests to cover the main code.

svrnm commented on charts/spegel/README.md:
```diff
@@ -69,6 +69,11 @@ Read the [getting started](https://spegel.dev/docs/getting-started/) guide to de
 | spegel.mirrorResolveTimeout | string | `"20ms"` | Max duration spent finding a mirror. |
 | spegel.mirroredRegistries | list | `[]` | Registries for which mirror configuration will be created. Empty means all registires will be mirrored. |
+| spegel.otel | object | `{"endpoint":"","insecure":false,"sampler":"parentbased_always_off","serviceName":"spegel"}` | OTEL tracing configuration. |
+| spegel.otel.endpoint | string | `""` | OTEL exporter endpoint (e.g., otel-collector:4318). |
```
It's not clear if this supports both OTLP over grpc and http or only one. It's good practice to support both.

svrnm reviewed (commented):
Hi, I provided an initial feedback from OTel best practices perspective, I'll do some more review later but overall this looks like a good start.

phillebaba commented on pkg/httpx/otel.go:
```diff
@@ -0,0 +1,25 @@
+// WrapTransport wraps an HTTP transport with OpenTelemetry instrumentation.
+func WrapTransport(name string, rt http.RoundTripper) http.RoundTripper {
+	if rt == nil {
+		rt = http.DefaultTransport
```
This is already the default in NewTransport.

dmathieu commented on internal/otelx/otelx_otel.go:
```diff
@@ -0,0 +1,168 @@
+		propagation.TraceContext{},
+		propagation.Baggage{},
+	)
+	if reflect.DeepEqual(otel.GetTextMapPropagator(), defaultPropagator) {
```
Using reflect here is a bit strange. Are you expecting folks to set propagators in other places? Is that a good pattern?

dmathieu commented on pkg/httpx/otel.go:
```diff
@@ -0,0 +1,25 @@
+		return otelhttp.NewTransport(rt)
+	}
+	return otelhttp.NewTransport(rt, otelhttp.WithSpanNameFormatter(func(operation string, _ *http.Request) string {
+		return name + " " + operation
```
I would recommend following the semantic convention for HTTP span names here.
https://github.com/open-telemetry/semantic-conventions/blob/main/docs/http/http-spans.md#name
